mdn/developer-portal
# 1513: Add backend logic for "Posts"-page search

This changeset adds backend-only support for searching the `title` and `description` fields of items which appear on the Posts (`/posts/`) page. (Specifically, those are `Article`s, `Video`s, `ExternalArticle`s and `ExternalVideo`s)

This will be followed by #1451 which will provide frontend access to this search behaviour. 

For now, one can test the search via URL manipulation (see below).

Notes:

* While the portal operates behind a CDN, the benefits of that are reduced substantially when search is involved -- cacheing paginated results in `/posts/` is at least potentially relevant to a greater proportion of visitors than cacheing results based on arbitrary search parameters. We will need to monitor load on the k8s cluster and adjust scaling rules appropriately. (Or we rebuild the Posts, Events and People pages entirely, in a worst case - eg to directly query an ES search backend from in-browser JS)

* Deep down, Wagtail's `postgres_search` backend does parameter quoting and escaping (see  See [Lexeme().as_sql()](https://github.com/wagtail/wagtail/blob/master/wagtail/contrib/postgres_search/query.py)), but opinions on whether what's currently implemented needs additional protections is very welcome.

(Related issue #1513 and parent #1473)

## How to test

- The code has been deployed to [stage](https://developer-portal.stage.mdn.mozit.cloud/)
- You can trigger a search by adding `search=some%20%terms` as a querystring. eg:
  - https://developer-portal.stage.mdn.mozit.cloud/posts/?search=css%20grid -- try this one in Safari too, to confirm the space works OK; try using pagination if it's offered
  - https://developer-portal.stage.mdn.mozit.cloud/posts/?search=covid-19

- It can also be mixed with existing filter selections from the sidebar:
  - https://developer-portal.stage.mdn.mozit.cloud/posts/?topic=css&search=flexbox 
  but please note that selecting a new topic and hitting "refine results" will lose any search params you have entered into the URL - that will be tackled in #1451

- Please confirm that pagination still works when you have search params in play (eg https://developer-portal.stage.mdn.mozit.cloud/posts/?search=firefox)
 
- We should leave the 'try to break it via the frontend' until #1451, but please do give the backend a grilling
